src/WebJobs.Script.WebHost/Security/KeyManagement/DefaultSecretsRepositoryFactory.cs

```diff
@@ -5,26 +5,68 @@
 using System.Collections.Generic;
 using System.IO;
 using System.Linq;
+using System.Threading.Tasks;
 using System.Web;
 using Microsoft.Azure.WebJobs.Host;
 using Microsoft.Azure.WebJobs.Script.Config;
+using Microsoft.WindowsAzure.Storage;
+using Microsoft.WindowsAzure.Storage.Blob;
 
 namespace Microsoft.Azure.WebJobs.Script.WebHost
 {
     public sealed class DefaultSecretsRepositoryFactory : ISecretsRepositoryFactory
     {
         public ISecretsRepository Create(ScriptSettingsManager settingsManager, WebHostSettings webHostSettings, ScriptHostConfiguration config)
         {
-            string secretStorageType = settingsManager.GetSetting(EnvironmentSettingNames.AzureWebJobsSecretStorageType);
-            string storageString = AmbientConnectionStringProvider.Instance.GetConnectionString(ConnectionStringNames.Storage);
-            if (secretStorageType != null && secretStorageType.Equals("Blob", StringComparison.OrdinalIgnoreCase) && storageString != null)
+            // 3 attempts to copy keys from file system to storage
+            for (int i = 0; i < 3; i++)
             {
+                string secretStorageType = settingsManager.GetSetting(EnvironmentSettingNames.AzureWebJobsSecretStorageType);
+                string storageString = AmbientConnectionStringProvider.Instance.GetConnectionString(ConnectionStringNames.Storage);
                 string siteSlotName = settingsManager.AzureWebsiteUniqueSlotName ?? config.HostConfig.HostId;
-                return new BlobStorageSecretsRepository(Path.Combine(webHostSettings.SecretsPath, "Sentinels"), storageString, siteSlotName);
+                if (secretStorageType != null && secretStorageType.Equals("Blob", StringComparison.OrdinalIgnoreCase) && storageString != null)
+                {
+                    return new BlobStorageSecretsRepository(Path.Combine(webHostSettings.SecretsPath, "Sentinels"), storageString, siteSlotName);
+                }
+                else
+                {
+                    if (string.IsNullOrEmpty(storageString))
+                    {
+                        return new FileSystemSecretsRepository(webHostSettings.SecretsPath);
+                    }
+                    try
+                    {
+                        // Change to blob storage repository
+                        CopyKeysFromFileSystemToBlobStorage(webHostSettings.SecretsPath, storageString, siteSlotName).Wait();
```

Comment:
This should be delayed and only done conditionally, as it will happen in the startup path if we're dealing with an HTTP request.

We should also avoid the synchronous `Wait` here, which is simpler to do if this is done lazily as all of the actual secret retrieval operations are async.